### PR TITLE
Provide stats for mixer client.

### LINKED ISSUE #132: Provide statistics
for
number of Check calls, 
number of cache hits
number of quota calls for each metric  
number of quota cache hits for each metric
number of report calls
number of remote report calls

### PR DESCRIPTION
**Pull request by JimmyCYJ**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #132 

**Special notes for your reviewer**:

**Release note**:

```release-note
Provide statistics for mixer client.
```

### CODE REVIEW

File: include/client.h [hunk 1]
```diff
@@ -48,6 +48,28 @@ struct MixerClientOptions {
   Environment env;
 };
 
+// The statistics recorded by mixerclient library.
+struct Statistics {
+  // Total number of check calls.
+  uint64_t total_check_calls;
+  // Total number of remote check hits.
+  uint64_t total_remote_check_calls;
+  // Total number of remote check calls that block inbound requests.
```
Comment (qiwzhang) [inline include/client.h]: that blocking origin requests.

File: include/client.h [hunk 2]
```diff
@@ -48,6 +48,28 @@ struct MixerClientOptions {
...
+  uint64_t total_blocking_remote_check_calls;
+
+  // Total number of quota calls for each metric.
```
Comment (qiwzhang) [inline include/client.h]: remove "for each metric"

File: include/client.h [hunk 3]
```diff
@@ -48,6 +48,28 @@ struct MixerClientOptions {
...
+  uint64_t total_quota_calls;
+  // Total number of remote quota calls for each metric.
+  uint64_t total_remote_quota_calls;
+  // Total number of remote quota calls that block inbound requests.
```
Comment (qiwzhang) [inline include/client.h]: that blocking origin requests

File: src/client_impl.h [hunk 4]
```diff
@@ -59,6 +60,18 @@ class MixerClientImpl : public MixerClient {
   std::string deduplication_id_base_;
   std::atomic<std::uint64_t> deduplication_id_;
 
+  // Atomic objects for recording statistics.
+  // check cache hit rate:
+  // 1 - total_blocking_remote_check_calls_ / total_check_calls_.
```
Comment (qiwzhang) [inline src/client_impl.h]: Your formula is for cache miss rate, not hit rate

File: src/report_batch.h [hunk 5]
```diff
@@ -32,12 +33,15 @@ class ReportBatch {
...
   virtual ~ReportBatch();
 
-  // Make batched report call
+  // Make batched report call.
   void Report(const ::istio::mixer::v1::Attributes& request);
 
   // Flush out batched reports.
   void Flush();
 
+  uint64_t total_report_calls() { return total_report_calls_; }
```
Comment (qiwzhang) [inline src/report_batch.h]: add "const"

File: src/report_batch.h [hunk 6]
```diff
@@ -16,11 +16,12 @@
 #ifndef MIXERCLIENT_REPORT_BATCH_H
 #define MIXERCLIENT_REPORT_BATCH_H
 
-#include <mutex>
-
 #include "include/client.h"
 #include "src/attribute_compressor.h"
 
+#include <atomic>
+#include <mutex>
```
Comment (qiwzhang) [inline src/report_batch.h]: do we need mutex?

File: src/client_impl_test.cc [hunk 7]
```diff
@@ -208,14 +248,24 @@ TEST_F(MixerClientImplTest, TestSuccessCheckAndQuota) {
...
     EXPECT_TRUE(done_status1.ok());
   }
   // Call count should be less than 4
   EXPECT_LE(call_counts, 3);
+  Statistics stat;
+  client_->GetStatistics(&stat);
+  // Less than 4 remote calls are made for prefetching, and they are
+  // non-blocking remote calls.
+  EXPECT_EQ(11, stat.total_check_calls);
+  EXPECT_GE(3, stat.total_remote_check_calls);
```
Comment (qiwzhang) [inline src/client_impl_test.cc]: remote_check_calls should be <=3, same as line 258

File: src/client_impl_test.cc [hunk 8]
```diff
@@ -208,14 +248,24 @@ TEST_F(MixerClientImplTest, TestSuccessCheckAndQuota) {
...
+  EXPECT_EQ(1, stat.total_blocking_remote_check_calls);
+  EXPECT_EQ(11, stat.total_quota_calls);
+  EXPECT_GE(3, stat.total_remote_quota_calls);
```
Comment (qiwzhang) [inline src/client_impl_test.cc]: same as line 264

File: src/report_batch.h (hunk 5 above)
Comment (JimmyCYJ) [reply src/report_batch.h]: Done.

File: src/report_batch.h (hunk 6 above)
Comment (JimmyCYJ) [reply src/report_batch.h]: Yes, we have std::mutex mutex_ at line 63.

File: src/client_impl_test.cc (hunk 8 above)
Comment (JimmyCYJ) [reply src/client_impl_test.cc]: Done.

File: src/client_impl_test.cc (hunk 7 above)
Comment (JimmyCYJ) [reply src/client_impl_test.cc]: Done.

File: src/client_impl.h (hunk 4 above)
Comment (JimmyCYJ) [reply src/client_impl.h]: I have changed it to calculating cache miss rate.

File: include/client.h (hunk 3 above)
Comment (JimmyCYJ) [reply include/client.h]: Comment is updated. Thanks!

File: include/client.h (hunk 2 above)
Comment (JimmyCYJ) [reply include/client.h]: Done. Thanks!

File: include/client.h (hunk 1 above)
Comment (JimmyCYJ) [reply include/client.h]: Comment is updated.

File: include/client.h [hunk 9]
```diff
@@ -48,6 +48,28 @@ struct MixerClientOptions {
   Environment env;
 };
 
+// The statistics recorded by mixerclient library.
+struct Statistics {
+  // Total number of check calls.
+  uint64_t total_check_calls;
+  // Total number of remote check hits.
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply include/client.h]: Done.

**Comment by JimmyCYJ** [comment]
Please take a look. Thanks!

**Pull request merged**